### ISSUE TITLE
Async data load without UI blocking

### ISSUE THREAD

**Issue by rokkolesa** [vaadin/vaadin-grid#2091]
### Description
When adding a `Grid` to a view with an `AbstractBackEndDataProvider` added, the view takes a longer time to display. This is due to it retrieving the data from the `DataProvider`, which can take a long time depending on the data source. All of this is done synchronously blocking the whole rendering of the view.

What I would like is to display the `Grid` with an overlay of some sort displaying a message (e.g. `Please wait, the data is loading`) with no rows, adding the rows when the actual load finishes and removing the message.

I tried to get around this by telling the `DataProvider` to not load any data initially - like "disabling" the `DataProvider` data fetch by returning a count of 0 and an empty `Stream` of data. Then I delegated the fetch to the next request using something like this:
```java
getElement().executeJs("")
            .then(json -> {
                <enable DataProvider load >
                grid.getDataProvider().refreshAll();
                <close the messafge before client response>
            });
```
The results were okay, the data was loaded only after the view rendered. However, the UI was blocked when the load was in progress and the user had to wait for it to finish.

I did some digging and found that the `DataCommunicator` always adds a function for execution before client response of the current request on a `DataChangeEvent` (method `requestFlush`.. it also adds it on initial attach). This means that there is no way of making the data load truly async as the UI is always blocked when this event is triggered otherwise an exception is thrown that the session is not locked.

Ideally the UI would not be blocked and the user would be able to interact with other components on the view as well as being able to navigate away from the view if the data load would take a long time, cancelling the data fetch. Is there a way to do this in the current implementation?

### Expected outcome
The data load should not block the view rendering.

### Actual outcome
Data load blocks the view rendering, not showing any changes until the data load is finished.

### Browsers Affected
All

### Versions:
- Vaadin / Flow / Grid version: `14.4.3` / `2.4.2` / `4.3.3`
- Java version: `11`
- OS version: `macOS Big Sur 11.0.1`
- Browser version (if applicable): `Chrome 87.0.4280.67`
- Application Server (if applicable): `WildFly 21.0.0`

**Comment by tomivirkki** [comment]
Hi @rokkolesa Sounds like something that shouldn't happen. Can you provide a minimal project with which the issue could be reproduced?

**Comment by rokkolesa** [comment]
Hi @tomivirkki. I attached a sample project. I used the bookstore starter and copied the relevant classes to `com.vaadin.starter.bookstore.ui.inventory.async` package. The important ones are `ProductBackendDataProvider`, `ProductAsyncGrid` and `InventoryAsyncView`. Other classes in this package are unchanged from their non-async counterparts (they only have different imports so I didn't have to make more abstract classes.. this is only a demo after all :)).

So the idea is that the `DataProvider` returns nothing until the next request, where the UI is fully built. However it still blocks the UI.. If I use another thread to call `DataProvider.refreshAll()`, then an exception is thrown if I don't wrap it in `ui.access` - which again blocks the UI.

Any idea to solve this issue would be welcome :)

[simple-app-grid-async.zip](https://github.com/vaadin/vaadin-grid/files/5665603/simple-app-grid-async.zip)

**Comment by tomivirkki** [comment]
Thanks for the example app. It demonstrates the issue well (the pending data request blocks the rest of the UI from updating). The problem is probably not Grid-specific but rather an issue with Vaadin in general. The same would happen with a simple button click listener if the roundtrip takes a long time to process.

```java
Button btn = new Button("Slow", clickEvent -> {
    try {
        Thread.sleep(5000);
    } catch (InterruptedException e) {
    }
});
add(btn);
```

Is there some built-in way to work around this @Legioth?

**Comment by rokkolesa** [comment]
I agree, but in a click listener implementation, I can always wrap the long-running task in another `Thread` and update the UI via `ui.access`. Something like this:
```java
Button btn = new Button("Slow", clickEvent -> {
    // show the loading indicator
    final UI ui = UI.getCurrent();
    new Thread(() ->
    {
        // long-running task
        try
        {
            Thread.sleep(5000);
        }
        catch (InterruptedException e)
        {
        }
        ui.access(() -> {
            // update the UI
            // hide the loading indicator
        });
    }).start();
});
add(btn);
```

However using a `DataProvider` I cannot do that as the long-running task is `dataProvider.refreshAll()` which triggers a `DataChangeEvent`.. `DataCommunicator` then handles this event with a listener added in `handleAttach`, resetting the whole state (method `reset`) which calls `requestFlush`.. this method however makes it mandatory for the UI to be locked as a `flushRequest` is added `beforeClientResponse`.

I get why this is implemented as it is.. to ensure the data flush happens only once - at the end of the request. But in such cases a conflict arises where I would not want to lock the UI for the data fetch, but it is mandatory to ensure it happens only once.. so another solution must be found..

It also seems that the majority of the problem is not in `Grid` but in `DataProvider`.. so feel free to move this issue where it belongs :)

**Comment by Legioth** [comment]
`DataProvider` is not designed to do asynchronous fetch operations. This is a tradeoff between optimal performance and avoiding to make an already complex part of the framework even more complex.

The new `DataView` concept that we've been adding might make it possible to add async functionality without too much compromise in the user-facing API. We would still need to find time to actually design a solution and implement it, so don't hold your breath...
